include/aspect/particle/integrator/interface.h

```diff
@@ -0,0 +1,227 @@
+#include <aspect/plugins.h>
+#include <aspect/global.h>
+
+#include <deal.II/base/parameter_handler.h>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Integrator
+    {
+      using namespace dealii;
+
+      /**
+       * An abstract class defining virtual methods for performing integration
+       * of particle paths through the simulation velocity field.
+       */
+      template <int dim>
+      class Interface
+      {
+        public:
+          /**
+           * Constructor.
+           */
+          Interface() {}
+
+          /**
+           * Destructor. Made virtual so that derived classes can be created
+           * and destroyed through pointers to the base class.
+           */
+          virtual ~Interface () {}
+
+          /**
+           * Perform an integration step of moving the particles of one cell
+           * by the specified timestep dt. Implementations of this function
+           * must update the particle location. Between calls to this function
+           * the velocity at the updated particle positions is evaluated and
+           * passed to integrate_step during the next call.
+           *
+           * @param [in] begin_particle An iterator to the first particle to be moved.
+           * @param [in] end_particle An iterator to the last particle to be moved.
+           * @param [in] old_velocities The velocities at t_n, i.e. before the
+           * particle movement.
+           * @param [in] velocities The velocities at t_{n+1}, i.e. after the
+           * particle movement.
+           * @param [in] dt The length of the integration timestep.
+           */
+          virtual
+          void
+          local_integrate_step(const typename std::multimap<LevelInd, Particle<dim> >::iterator &begin_particle,
+                               const typename std::multimap<LevelInd, Particle<dim> >::iterator &end_particle,
+                               const std::vector<Tensor<1,dim> > &old_velocities,
+                               const std::vector<Tensor<1,dim> > &velocities,
+                               const double dt) = 0;
+
+          /**
+           * This function is called at the end of every integration step.
+           * In case of multi-step integrators is signals the beginning of a
+           * new integration step.
+           */
+          virtual void advance_step();
+
+          /**
+           * @return This function returns true if another integration step is required
+           * by the integrator. Its default implementation returns false.
+           */
+          virtual bool continue_integration() const;
+
+          /**
+           * Return data length of the integration related data required for
+           * communication in terms of number of doubles.
+           *
+           * @return The number of doubles required to store the relevant
+           * integrator data for one particle.
+           */
+          virtual unsigned int data_length() const;
```

Comment:
Could you foresee a case where an integrator wants to store anything other than `double` values? In that case, counting in bytes may be more useful. But it may be unnecessary to think about this.

In general, I think the function documentation lacks context. Maybe a sentence of the sort "When data about particles is transported from one processor to another, or stored on disk for snapshots, integrators get the chance to store whatever information they need with each particle. This function returns how many pieces of additional information a concrete integrator class needs to store with each particle."